Allow usage of new MQTT 5 implementation

underhood opened the pull request:
##### Summary

Allows usage of the new MQTT 5 custom client. https://github.com/netdata/netdata/issues/12481
By default disabled for now until it is polished and stability is tested well enough.

Main benefit -> removed max publish size limit. The published data don't all have to fit into a fixed size buffer anymore.
As a result also memory usage is reduced.

Currently implements only the bits and pieces of MQTT5 which we actually use. In future more will be implemented (topic aliases might be useful for us).

Following PRs need to implement smarter query thread that actually throttles when there is too much data pending.

##### Test Plan
Absolutely most important piece to test is this by accident doesn't break the current MQTT by setting `mqtt5 = no`.

Then you can go ahead and have fun with this new implementation

```
netdata.conf
[cloud]
    mqtt5 = yes
```


##### Additional Information
Some additional info can be found here:
Fixes #12481

<details> <summary>For users: How does this change affect me?</summary>
  <!--
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ACLK (cloud connection)
- Can they see the change or is it an under the hood? If they can see it, where? Under the hood
- How is the user impacted by the change? Better Agent<->Cloud performance especially for bigger setups. Lowered memory consumption.
- What are there any benefits of the change? lower mem consumption, removed message size limit
-->
</details>

stelfrag commented:
Some errors
```
2022-05-13 17:51:27: netdata ERROR : ACLK_Query_11 : [mqtt_wss] E: mqtt_client: mqtt_ng_publish failed to generate message due to insufficient buffer space (line 913)
2022-05-13 17:51:27: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_puback failed to generate message due to insufficient buffer space (line 1066)
2022-05-13 17:51:27: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: Error generating PUBACK reply for PUBLISH
2022-05-13 17:51:27: netdata ERROR : ACLK_Query_39 : [mqtt_wss] E: mqtt_client: mqtt_ng_publish failed to generate message due to insufficient buffer space (line 913)
2022-05-13 17:51:27: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_puback failed to generate message due to insufficient buffer space (line 1066)
2022-05-13 17:51:27: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: Error generating PUBACK reply for PUBLISH
```

stelfrag commented:
Also 

```
2022-05-13 17:49:14: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_ping failed to generate message due to insufficient buffer space (line 1095)
2022-05-13 17:49:14: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_ping failed to generate message due to insufficient buffer space (line 1095)
2022-05-13 17:49:14: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_ping failed to generate message due to insufficient buffer space (line 1095)
2022-05-13 17:49:14: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_ping failed to generate message due to insufficient buffer space (line 1095)
2022-05-13 17:49:14: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_ping failed to generate message due to insufficient buffer space (line 1095)
2022-05-13 17:49:14: netdata ERROR : ACLK_Main : [mqtt_wss] E: mqtt_client: mqtt_ng_ping failed to generate message due to insufficient buffer space (line 1095
```

stelfrag commented:
```
#1  0x00007fb5bda23537 in __GI_abort () at abort.c:79
#2  0x00007fb5bda7c768 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7fb5bdb8ae2d "%s\n") at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007fb5bda83a5a in malloc_printerr (str=str@entry=0x7fb5bdb8d230 "double free or corruption (!prev)") at malloc.c:5347
#4  0x00007fb5bda8509c in _int_free (av=0x7fb5bdbbcb80 <main_arena>, p=0x563104d59870, have_lock=<optimized out>) at malloc.c:4317
#5  0x0000562efb1fcedc in freez (ptr=<optimized out>) at libnetdata/libnetdata.c:166
#6  0x0000562efb3ea567 in buffer_frag_free_data (frag=frag@entry=0x562fe6b01083) at mqtt_websockets/src/mqtt_ng.c:410
#7  0x0000562efb3ea6a5 in buffer_garbage_collect (client=client@entry=0x562fe6a1cf60) at mqtt_websockets/src/mqtt_ng.c:432
#8  0x0000562efb3eb8dc in mqtt_ng_publish (client=0x562fe6a1cf60, 
    topic=topic@entry=0x5630af7a8230 "/svc/agent-data-ctrl/c1611d96-4936-41d0-bdd1-b3c820c09c12a839579d-c91b-4dd3-bc07-96bc70b80ab4", topic_free=topic_free@entry=0x0, 
    msg=msg@entry=0x56312b726d80, msg_free=msg_free@entry=0x0, msg_len=773, publish_flags=2 '\002', packet_id=0x7fb58dc609ce) at mqtt_websockets/src/mqtt_ng.c:913
#9  0x0000562efb3e8a08 in mqtt_wss_publish5 (client=client@entry=0x562efd04d500, 
    topic=topic@entry=0x5630af7a8230 "/svc/agent-data-ctrl/c1611d96-4936-41d0-bdd1-b3c820c09c12a839579d-c91b-4dd3-bc07-96bc70b80ab4", topic_free=topic_free@entry=0x0, 
    msg=msg@entry=0x56312b726d80, msg_free=msg_free@entry=0x0, msg_len=<optimized out>, publish_flags=1 '\001', packet_id=0x7fb58dc609ce)
    at mqtt_websockets/src/mqtt_wss_client.c:1327
#10 0x0000562efb3e33e8 in aclk_send_message_with_bin_payload (client=client@entry=0x562efd04d500, msg=msg@entry=0x5630b185a1c0, 
    topic=topic@entry=0x5630af7a8230 "/svc/agent-data-ctrl/c1611d96-4936-41d0-bdd1-b3c820c09c12a839579d-c91b-4dd3-bc07-96bc70b80ab4", 
    payload=payload@entry=0x5630c7604480, payload_len=payload_len@entry=471) at aclk/aclk_tx_msgs.c:157
#11 0x0000562efb3e3a18 in aclk_http_msg_v2 (client=0x562efd04d500, 
    topic=0x5630af7a8230 "/svc/agent-data-ctrl/c1611d96-4936-41d0-bdd1-b3c820c09c12a839579d-c91b-4dd3-bc07-96bc70b80ab4", 
    msg_id=0x5630fbdc4a00 "55ce1b60-b274-11ec-99c7-ff587cb3880e/f3ad1480-8605-4a87-bf10-7bda0d443d43/data/1652774933063393625", t_exec=t_exec@entry=1025, 
    created=1652774933262040, http_code=404, 
    payload=0x5630c7604480 "HTTP/1.1 404 Not Found\r\nConnection: close\r\nServer: Netdata Embedded HTTP Server v1.34.0-211-g87f67e38c\r\nAccess-Control-Allow-Origin: *\r\nAccess-Control-Allow-Credentials: true\r\nContent-Type: text/plain"..., payload_len=471) at aclk/aclk_tx_msgs.c:371
#12 0x0000562efb3e02fc in http_api_v2 (query_thr=0x562fe6b33798, query=0x56313bd806c0) at aclk/aclk_query.c:249
#13 0x0000562efb3e0879 in aclk_query_process_msg (query=<optimized out>, query_thr=<optimized out>) at aclk/aclk_query.c:349
#14 aclk_query_process_msgs (query_thr=query_thr@entry=0x562fe6b33798) at aclk/aclk_query.c:372
#15 0x0000562efb3e09b1 in aclk_query_main_thread (ptr=0x562fe6b33798) at aclk/aclk_query.c:394
#16 0x0000562efb209737 in thread_start (ptr=<optimized out>) at libnetdata/threads/threads.c:185
#17 0x00007fb5bdef8ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
#18 0x00007fb5bdafbdef in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

stelfrag commented:
Some compilation warnings
```
aclk/aclk.c:59:9: warning: 'RSA_free' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
aclk/aclk.c:78:5: warning: 'PEM_read_bio_RSAPrivateKey' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
aclk/aclk.c:400:36: warning: passing argument 2 of 'mqtt_wss_subscribe' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
aclk/aclk.c:408:40: warning: passing argument 2 of 'mqtt_wss_subscribe' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
aclk/aclk.c:1098:28: warning: passing argument 1 of 'freez' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
aclk/aclk.c:1131:36: warning: passing argument 1 of 'freez' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
aclk/aclk_tx_msgs.c:157:35: warning: passing argument 2 of 'mqtt_wss_publish5' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
aclk/aclk_tx_msgs.c:157:71: warning: passing argument 4 of 'mqtt_wss_publish5' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
aclk/aclk_otp.c:161:5: warning: 'RSA_private_decrypt' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
```

MrZammler commented on aclk/aclk.c:
```diff
@@ -1096,39 +1112,52 @@ void aclk_send_node_instances()
             create_query = aclk_query_new(REGISTER_NODE);
+            node_instance_creation_t node_instance_creation = {
+                .hops = list->hops,
+                .hostname = list->hostname,
```
Hi Timo! Do we need to free `list->hostname` ?

underhood commented on aclk/aclk_tx_msgs.c:
```diff
@@ -149,17 +153,22 @@ static int aclk_send_message_with_bin_payload(mqtt_wss_client client, json_objec
+        if (rc == MQTT_WSS_ERR_BLOCK_TIMEOUT) {
+            error("Timeout sending binpacked message");
+            freez(full_msg);
+            return 503;
```
Hi, thanks that is good suggestion in this PR i will change it in places this PR touch.

I will make another PR for another lines of code that have the same issue.

underhood commented:
> .hostname = list->hostname,
> Member

removed warnings that are not already in master for example OpenSSL 3 warnings:
   - there is issue for that, the Challenge/Response crypto code which is perhaps the most ancient bit of ACLK code in existence  (apart from claiming script) uses functions that are deprecated in OpenSSL3. I will have to rewrite that with new API

underhood commented:
after addressing reviewer comment I also rebased on latest master (to avoid possible problems that have been fixed on master meanwhile)
there was no conflicts during rebase

thiagoftsm commented:
Hello @underhood ,

I tested with option `mqtt5 = no` and everything works fine and the messages given are ok:

```sh
2022-06-03 21:30:39: netdata INFO  : ACLK_Main : MQTT App Layer `disconnect` message sent successfully
2022-06-03 21:30:39: netdata INFO  : ACLK_Main : Attempting to gracefully shutdown the MQTT/WSS connection
2022-06-03 21:31:04: netdata INFO  : ACLK_Main : [mqtt_wss] I: Going to connect using MQTT-C (forked) as MQTT implementation
```

On the other hand, when I tried to use the latest version (`mqtt5 = yes`) I received the following messages with an error:

```sh
2022-06-03 21:16:24: netdata INFO  : ACLK_Main : [mqtt_wss] I: Going to connect using internal MQTT 5 implementation
2022-06-03 21:16:24: netdata ERROR : ACLK_Main : [mqtt_wss] W: client_id provided is longer than 23 bytes, server might not allow that [MQTT-3.1.3-5]
2022-06-03 21:16:25: netdata INFO  : ACLK_Main : [mqtt_wss] I: mqtt_client: MQTT Connection Accepted By Server
```

The presence of an error did not influence the final result, I had the cloud working as expected, but is this expected?

underhood commented:
@thiagoftsm `W: client_id provided is longer than 23 bytes, server might not allow that [MQTT-3.1.3-5]`

`[MQTT-3.1.3-5]` this is reference for MQTT specification and it states that server don't have to support client_id longer than 23 bytes. Most of servers do that. But if server will not accept such connection it is OK. Therefore I print that as a warning
